### PR TITLE
Added fix for permissions, owner to pgsql directory

### LINKED ISSUE #64: Permissions 0777 set on some directories within Dockerfile
As discussed in openshift/mysql/pull/94 and mysql/pull/92#issuecomment-139270899.

Fwiw "Reorder Dockerfile for better use of cache" commit message did not help me
to understand the reasoning for the `chmod a+rwx ..` change...

### PR DESCRIPTION
**Pull request by Akasurde**
Fixes #64 
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

### CODE REVIEW

**Comment by mfojtik** [comment]
@praiskup review pls

**Comment by mfojtik** [comment]
[test]

**Comment by praiskup** [comment]
I dislike the 'a+rwx' a lot -- this sounds like step forward, but reasoning
why this change is needed and why exactly `postgres.0`, etc.. is needed.

**Comment by mfojtik** [comment]
@praiskup the chgroup 0 is needed for this image to run in OpenShift IMHO (the user is random, but the group is always 0)

**Comment by mfojtik** [comment]
@praiskup in other words, restricting write to the root group is better than having it world writable ;-)

**Comment by praiskup** [comment]
I see, but mysql has e.g. in-code comment:
https://github.com/openshift/mysql/blob/master/5.6/root/usr/libexec/container-setup#L53

File: 9.2/Dockerfile [hunk 1]
```diff
@@ -32,13 +32,13 @@ RUN rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 && \
     yum -y --setopt=tsflags=nodocs install nss_wrapper && \
     yum clean all && \
     localedef -f UTF-8 -i en_US en_US.UTF-8 && \
-    mkdir -p /var/lib/pgsql/data && chown postgres.postgres /var/lib/pgsql/data && \
+    mkdir -p /var/lib/pgsql/data && chown -R postgres.0 /var/lib/pgsql && \
     test "$(id postgres)" = "uid=26(postgres) gid=26(postgres) groups=26(postgres)" && \
     # Loosen permission bits to avoid problems running container with arbitrary UID
     chmod -R a+rwx /var/run/postgresql
```
Comment (bparees) [inline 9.2/Dockerfile]: why not change this one too?

**Comment by bparees** [comment]
two main comments:
1) please update the Dockerfile.rhel7 files too
2) we now have a fix_permissions script we use for stuff like this, it would be better to use it so we can have some consistency in how permissions get set: 
https://github.com/openshift/sti-base/blob/master/bin/fix-permissions
(you'll have to copy that script into this repo since this image isn't based on the sti-base image)

once that script is in the image, you can invoke it with the directory to "fix" as the parameter.

**Comment by Akasurde** [comment]
OK, I will do this and revert back.

**Comment by Akasurde** [comment]
Ping

**Comment by bparees** [comment]
lgtm, but can you please squash your commits?  thanks.
[test]

**Comment by praiskup** [comment]
@bparees wrote:
> lgtm, but can you please squash your commits? thanks.

FWIW:  Does not look good to me.

We should probably move the `fix-permissions` script at least into
`root/usr/libexec`, or if it will stay in $PATH (and not removed before the
image is squashed) could we name it like `artificial-uid-perm-fix`?

**Comment by Akasurde** [comment]
@praiskup We can move `fix-permissions` to `/usr/libexec`  directory and run it from there.

**Comment by praiskup** [comment]
+1 if you think about new `libexec` subdir in https://github.com/openshift/postgresql/tree/master/9.4/root/usr

**Comment by praiskup** [comment]
Is this fine to [test]?  I would quote variables in fix-permissions script, but not
a problem with actual hard-wired calls from Dockerfiles.

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
`failed: mkdir: cannot create directory '/var/lib/pgsql/data/userdata': Permission denied`

ironically when you actually run as the "postgres" user, you are not a member of the "root" group because uid 26 can actually be found in /etc/passwd with a group of postgres.

the easiest fix is probably to set the user ownership to postgres, in fix-permissions.  That way the files will be writable by either the postgres user, or anyone else (because anyone else will be a member of the root group)

**Comment by Akasurde** [comment]
@bparees 

Adding following line in `fix-permissions` fixes issue

`chown -R postgres $1`

Let me know if this is OK?

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
I don't see the change, did you push it?

**Comment by praiskup** [comment]
@Akasurde Note the missing quotation around $1.

**Comment by bparees** [comment]
@mnagy is this PR necessary to fix replication, or is this just clean up?  I don't want to merge anything we don't need.

File: 9.2/root/usr/libexec/fix-permissions [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+# Fix permissions on the given directory to allow group read/write of
+# regular files and execute of directories.
+chown -R postgres "$1"
+chgrp -R 0 "$1"
+chmod -R g+rw "$1"
+find $1 -type d -exec chmod g+x {} +
```
Comment (mnagy) [inline 9.2/root/usr/libexec/fix-permissions]: Quotes around $1

Comment (mnagy) [inline 9.2/root/usr/libexec/fix-permissions]: please add

File: 9.4/root/usr/libexec/fix-permissions [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+# Fix permissions on the given directory to allow group read/write of
+# regular files and execute of directories.
+chown -R postgres "$1"
+chgrp -R 0 "$1"
+chmod -R g+rw "$1"
+find $1 -type d -exec chmod g+x {} +
```
Comment (mnagy) [inline 9.4/root/usr/libexec/fix-permissions]: same here

**Comment by mnagy** [comment]
Shouldn't we call the fix-permission script from the entrypoint as well?

**Comment by Akasurde** [comment]
@mnagy  Do we need to call `fix-permissions` entrypoint as well ?

**Comment by praiskup** [comment]
I don't think so.  You even do not have usually permissions to change ownership of file.

**Comment by bparees** [comment]
not needed for replication fixing, deferring this until we get replication straightened out.

**Comment by Akasurde** [comment]
@bparees ping

**Comment by mnagy** [comment]
I'd like to note that it would be a very good idea to test this PR with OS extended tests we now have for replication.

**Comment by bparees** [comment]
[test]

File: 9.4/root/usr/libexec/fix-permissions [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+# Fix permissions on the given directory to allow group read/write of
+# regular files and execute of directories.
+chown -R postgres "$1"
+chgrp -R 0 "$1"
+chmod -R g+rw "$1"
+find "$1" -type d -exec chmod g+x {} +
```
Comment (mnagy) [inline 9.4/root/usr/libexec/fix-permissions]: Are these commands allowed to fail? If not, we should add `set -e`. We should also make sure $1 is defined. Together with what we had previously, `set -eu` is ideal.

**Comment by bparees** [comment]
@mnagy is testing this now for replication.  going forward we're going to ask that the SCL team start running the openshift extended image tests(which include replication) on these PRs (the PR tester does a more basic test, it doesn't test replication scenarios for example).

we can discuss how best to approach this on tuesday.

**Comment by mnagy** [comment]
FYI, verified these changes and tests passed.

File: 9.4/root/usr/libexec/fix-permissions (hunk 4 above)
Comment (bparees) [inline 9.4/root/usr/libexec/fix-permissions]: it's not in our sti-base script either:
https://github.com/openshift/sti-base/blob/master/bin/fix-permissions

updating this script here and elsewhere can be a follow up.

**Pull request merged**